---
title: "CHAI: consensus Clustering tHrough similArIty matrIx integratIon for single cell type identification"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{chai_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# Introduction
CHAI (consensus Clustering tHrough similArIty matrIx integratIon for single cell type identification) is a consensus clustering framework that offers two methods for consensus clustering: Average Similarity (AvgSim) and Similarity Network Fusion (SNF).
This vignette demonstrates a standard workflow for running both AvgSim and SNF for the purpose of consensus clustering. 

# Install Dependencies
All CRAN related dependencies are installed with the installion of CHAI. However you will need to install some other dependencies independently from BioConductor and GitHub. Below is how to do so for each package.
```{r setup, eval=FALSE}
# Install scSHC
devtools::install_github("igrabski/sc-SHC")
# Install RaceID
devtools::install_github("dgrun/RaceID3_StemID2_package")
# Install SC3
if (!require("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}
BiocManager::install("SC3")
# Install SingleCellExperiment
BiocManager::install("SingleCellExperiment")

remotes::install_github("corceslab/CHOIR", ref="main", repos = BiocManager::repositories(), upgrade = "never")
```
Now, all packages should be installed.
```{r, message=FALSE}
library(chai)
library(scSHC)
library(RaceID)
library(SC3)
library(SingleCellExperiment)
library(CHOIR)
```
# Data Loading and Preprocessing
CHAI accepts a SingleCellExperiment object as input for running all clustering algorithms. Creating a SingleCellExperiment object for single cell RNA-seq data is well documented, here is an example for creating one using a single cell matrix, where rows are genes and columns are cells. 
Be sure to load the counts and logcounts into the single cell experiment object. 

For test data, we provide the Mouse_1 dataset from the widely used Baron datasets. This is accessible within the CHAI package. 

```{r, eval=FALSE}
# Load data
data("baron_mouse_1")

# Create SingleCellExperiment object
sce <- SingleCellExperiment(assays = list(counts = as.matrix(baron_mouse_1)))
# Add logcounts 
sce <- scuttle::logNormCounts(sce)

```

# Run Clustering Algorithms
The CHAI workflow currently supports 6 state-of-the-art clustering methods: Seurat, scSHC, Spectrum, SC3, RaceID and CHOIR. For Seurat we run two clustering algorithms, Louvain and SLC. To run all algorithms, we provide a wrapper function that takes a SingleCellExperiment object as input. 

Each clustering assignment is added to the colData() of the SingleCellExperiment object.

For more advanced parameter options, please see the documentation of the ```get_clust_assignments``` function. 

NOTE: If you are using a Windows machine, please set n_cores to 1. 


```{r, eval=FALSE}
# Run All Clustering Algorithms
sce <- get_clust_assignments(sce, n_cores = 12, svm_max = 200, max_k = 15)
```

You also have the ability to run individual clustering algorithms. Simply call ```sce <- {algorithm_name}_assign_func(sce)```. Here is an example for Seurat.
```{r, eval=FALSE}
# Run Just Seurat Clustering Algorithm. 
sce <- seurat_assign_func(sce)
```

# Create Binary Similarity Matrices

Both AvgSim and SNF consensus clustering are based on binary similarity matrices. In brief, we create a similarity matrix for each cluster assignment with the following rules:

1. If two cells are assigned the same cluster, assign a score of 1. 
2. If two cells are not assigned the same cluster, assign a score of zero. 

Therefore, we are able to create a binary similarity matrix for each clustering algorithm in the CHAI workflow. We store each matrix in a list format, which makes them easier to access for downstream functions. 

The function to create the list of similarity matrices takes as parameter the original SingleCellExperiment object. We also recommend checking the length of similarity_matrix_list to make sure that the number of desired algorithms are present in the list.
```{r, eval=FALSE}
# Generate list of binary similarity matrices 
similarity_matrix_list = create_matrix_list(sce)

length(similarity_matrix_list)
```

# Run AvgSim or SNF

To run AvgSim, we must create an average ranking matrix of each binary similarity matrix in the similarity_matrix_list.

```{r,eval=FALSE}
# Generate list of binary similarity matrices 
AvgSim_matrix <- create_average_matrix(similarity_matrix_list)
```

To run SNF, we must call the SNF function on the similarity_matrix_list. 

```{r, eval=FALSE}
# Generate list of binary similarity matrices 
snf_matrix <- create_snf_matrix(similarity_matrix_list)
```

# Run Spectral Clustering for Consensus Clustering Assignments 

CHAI supports running Spectral Clustering on either the AvgSim_matrix or the snf_matrix. Spectral Clustering requires the specification of a "K" partitioning of the graph, which will be the number of cell clusters in this case. 

If a user already knows the true number of clusters in the dataset, they can simply run Spectral Clustering on either matrix. We use the AvgSim_matrix as an example here, but it is interchangable with the SNF_matrix in the case of this function. 
```{r, eval=FALSE}
# Run Spectral clustering on AvgSim_matrix. To run on snf_matrix, replace AvgSim_matrix with snf_matrix in this function. 
true_k <- 9
AvgSim_matrix_clusters <- spectral_clustering(AvgSim_matrix, true_k)
```

If the true number of clusters is not known, users can determine the "best_k" for Spectral Clustering by evaluating the optimal silhouette score. The CHAI workflow provides a method to do so. 
It is important to do separate evaluations for best_k on the AvgSim_matrix and SNF_matrix, as the data distributions are different, therefore the optimal partioning may be different as well.
```{r,eval=FALSE}

# Determine best K through silhouette score. 
AvgSim_best_k <- calc_silhouette_scores(AvgSim_matrix, 15)

# To run on snf_matrix, replace AvgSim_matrix with SNF_matrix in this function, as well as either the true number of clusters in the dataset or evaluate for best K. 

AvgSim_clusters <- spectral_clustering(AvgSim_matrix, AvgSim_best_k)
```

# Evaluation

If known ground_truth is available, CHAI allows you to create a table with Adjusted Rand Index (ARI) and Normalized Mutual Information (NMI). 
CHAI calculates the ARI and NMI for each clustering assignment in ```colData(sce)```.

We provide the ground truth labels for the example dataset by loading ```data("baron_mouse_1_ground_truth")```, stored in the column ```clust_assign```. 

NOTE: When adding the CHAI assignments to the SingleCellExperiment object, it is crucial to use the word ```"_assign"``` in the column as we have done below, or else the ```evaluation_table``` function will not recognize it as a clustering assignment column. 

```{r, eval=FALSE}
data("baron_mouse_1_ground_truth")
# First, add CHAI_AvgSim_assign to the colData of the SingleCellExperiment object. 
colData(sce)$CHAI_AvgSim_assign <- AvgSim_clusters@.Data

# Create evaluation table. 
evaluation_table <- evaluation_table(sce, baron_mouse_1_ground_truth$clust_assign)

```

# Session Info 

```{r}
sessionInfo()
```

# TODO: Add Visualization, Quick Start, and References.  